specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+           1. Check that `response.signatureEnvelope` can be verified using the public key and signing algorithm specified in the signing certificate, which is embedded as part of certificate chain in `response.signatureEnvelope` . This step does not include certificate chain validation (certificate chain leads to a trusted root configure in Notation), or revocation check.
+           1. Check that the certificate chain in `response.signatureEnvelope` confirm to [Certificate Requirements].
+        1. Generate a signature manifest for the given signature envelope, and append `response.annotations` to manifest annotations.
+    1. Else if plugin supports capability `SIGNATURE_GENERATOR` *(covered in previous section)*
+    1. Return an error
+
+**generate-envelope**
+
+*Request*
+All request attributes are required.
+
+```json
+{
+  "contract-version" : <major-version.minor-version>,
+  
+  // Complete key definition from /notation/config.JSON /signingKeys/keys with matching key name
+  "key" : <key definition>, 
+
+  "payload" : <Base64 encoded payload to be signed>,
+  
+  // The type of payload - currently a descriptor
+  "payloadType" : "application/vnd.oci.descriptor.v1+json",
+  
+  // The expected response signature envelope
+  "signatureEnvelopeType" : "application/vnd.cncf.notary.v2.jws.v1"
+}
+```
+
+*signatureEnvelopeType* - defines the type of signature envelope expected from the plugin. As  Notation clients need to be updated in order to parse and verify new signature formats, the default signature format can only be changed with new major version releases of Notation. Users however can opt into using an updated signature format supported by Notation, by passing an optional parameter.
+e.g. `notation sign $IMAGE --key {key-name} --signatureFormat {some-new-format}`
+
+*Response*
+All response attributes are required.
+
+
+```json
+{
+   "signatureEnvelope": <Base64 encoded signature envelope>,
+   "signatureEnvelopeType" : "application/vnd.cncf.notary.v2.jws.v1"
+   
+  // Annotations to be appended to Signature Manifest annotations
+  "annotations" : {
+    "key1" : "value1",
+    "key2" : "value2"
+  }
+}
+```
+
+*Error codes*
+
+* VALIDATION_ERROR - Any of the required request fields was empty, or a value was malformed/invalid. Includes condition where the key referenced by `keyId` was not found, payload type or signature envelope type in the request is unsupported by the plugin.
+* UNSUPPORTED_CONTRACT_VERSION - The contract version used in the request is unsupported.
+* ACCESS_DENIED - Authentication/authorization error to use given key.
+* TIMEOUT - The operation to generate signature timed out and can be retried by Notation.
+* THROTTLED - The operation to generate signature was throttles and can be retried by Notation.
+* ERROR - Any general error that does not fall into previous error categories.
+
+## Verification extensibility
+
+**TBD**
```

Comment:
Having a issue to track this before the next release would be good. Similarly for the **Threat Model** as well.